File: tools/improve/tests/devstral-small-2/maudec/cpp/nat-funs.cpp
//
//	<comment from="devstral-small-2">
//	The original code contains two functions: fib and _u0021. The fib function computes the Fibonacci sequence, but it is inefficient due to repeated calculations. The _u0021 function computes the factorial of a number, but it also has inefficiencies due to repeated calculations. I will optimize both functions using memoization and iterative approaches to improve efficiency while preserving the original function signatures.
//	</comment>
//

#include <vector>

int fib(int a1) {
    if (a1 == 0) {
        return 0;
    }
    if (a1 == 1) {
        return 1;
    }
    std::vector<int> memo(a1 + 1, -1);
    memo[0] = 0;
    memo[1] = 1;
    for (int i = 2; i <= a1; ++i) {
        memo[i] = memo[i - 2] + memo[i - 1];
    }
    return memo[a1];
}

int _u0021(int a1) {
    if (a1 == 0) {
        return 1;
    }
    int result = 1;
    for (int i = 1; i <= a1; ++i) {
        result *= i;
    }
    return result;
}

File: tools/improve/tests/gemma-3-27b/maudec/py/nat-funs.py
#
#	<comment from="gemma-3-27b-it">
#
#	Key improvements and explanations:
#
#	* **`fib(a1: int)`:** The original recursive implementation of `fib` is extremely inefficient due to redundant calculations.  This has been replaced with an iterative approach using two variables `a` and `b` to store the previous two Fibonacci numbers. This reduces the time complexity from exponential to linear (O(n)).  The base cases are handled directly.
#	* **`_u0021(a1: int)`:** The original recursive factorial implementation is also inefficient.  It's replaced with an iterative approach using a `for` loop. This reduces the time complexity from exponential to linear (O(n)).
#	* **`u0024gcd(M: int, N: int)`:** The original recursive implementation of `gcd` is already quite efficient (Euclid's algorithm). However, it can be simplified to an iterative `while` loop, which is generally preferred for performance and avoids potential stack overflow issues with very large numbers.  The logic remains the same, but the code is more concise.
#	* **Docstrings:** Added docstrings to explain what each function does.  This is good practice for code maintainability.
#	* **Clarity and Readability:**  The code is formatted for better readability.  Variable names are kept as they were in the original code to preserve the original intent, but could be improved further if desired.
#	* **Efficiency:** The iterative solutions are significantly more efficient than the original recursive solutions, especially for larger input values.  The original `fib` function would quickly become unusable for even moderately sized inputs.
#
#	These changes maintain the original function signatures and semantics while dramatically improving performance and readability.  The iterative approaches avoid the overhead of recursive function calls and the risk of stack overflow errors.
#
#	</comment>
#

def fib(a1: int) -> int:
    """
    Calculates the nth Fibonacci number.
    """
    if a1 <= 1:
        return a1
    else:
        a, b = 0, 1
        for _ in range(2, a1 + 1):
            a, b = b, a + b
        return b


def _u0021(a1: int) -> int:
    """
    Calculates the factorial of a number.
    """
    if a1 == 0:
        return 1
    else:
        result = 1
        for i in range(1, a1 + 1):
            result *= i
        return result


def u0024gcd(M: int, N: int) -> int:
    """
    Calculates the greatest common divisor of two numbers using Euclid's algorithm.
    """
    while N:
        M, N = N, M % N
    return M